### PR TITLE
King is back 

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Only for distress and hunt

The king is summoned by the queen through a King Pod, which costs 2500 psy points. 10 minutes to hatch

New Abilities:
Nightfall: All lights are shutdown around king, for 5 seconds. Flares are immune
Gravity crush: Medium wind up after which a 3x3 area is heavily damaged, good for cades
Hive summon: After a wind-up with telegraphs the entire hive is teleported to the king

## Why It's Good For The Game

Closes https://github.com/tgstation/TerraGov-Marine-Corps/issues/6402

Xeno content?

## Changelog
TiviPlus, revived by Bravemole
:cl:
add: King is back
/:cl:

### CODE REVIEW

**Comment by ilikesaying** [comment]
back from buying spess milk

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 1]
```diff
@@ -0,0 +1,135 @@
+// ***************************************
+// *********** Nightfall
+// ***************************************
+
+/datum/action/xeno_action/activable/nightfall
+	name = "Nightfall"
+	action_icon_state = "nightfall"
+	ability_name = "Nightfall"
+	mechanics_text = "Shut down all electrical lights for 5 seconds"
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:9]:
```suggestion
	mechanics_text = "Shut down all electrical lights for five seconds."
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	var/list/turfs = RANGE_TURFS(2, A)
+	for(var/turf/targetted AS in turfs)
+		targetted.add_filter("crushblur", 1, list("type"="radial_blur", "size" = 0.3))
+	if(!do_after(owner, 2 SECONDS, FALSE, owner, BUSY_ICON_DANGER))
+		for(var/turf/targetted AS in turfs)
+			targetted.remove_filter("crushblur")
+		return fail_activate()
+	succeed_activate()
+	add_cooldown()
+	A.visible_message("<span class='warning'> [A] collapses inward as it's gravity suddenly increases!</span>")
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:69]:
```suggestion
	A.visible_message("<span class='warning'>[A] collapses inward as it's gravity suddenly increases!</span>")
```

**Comment by Losenis** [comment]
Bruh you straight up get over 2k points almost by default save for lowpop

**Comment by lKiseki** [comment]
One day, our father left to the convenience store and never came back.

But today, he returns. Hopefully with the space milk.

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+/datum/action/xeno_action/activable/nightfall/on_cooldown_finish()
+	to_chat(owner, "<span class='notice'>We gather enough mental strength to shut down lights again.</span>")
+	return ..()
+
+/datum/action/xeno_action/activable/nightfall/use_ability()
+	playsound(owner, 'sound/magic/nightfall.ogg', 50, 1)
+
+	for(var/atom/light AS in GLOB.lights)
+		light.turn_light(null, FALSE, duration, TRUE, TRUE, get_turf(owner), range)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:25]: this is unperformant and should use a range check instead

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	var/mob/living/carbon/xenomorph/X = owner
+
+	log_game("[key_name(owner)] has begun summoning hive in [AREACOORD(owner)]")
+	xeno_message("King: \The [owner] has begun a psychic summon in <b>[get_area(owner)]</b>!", "xenoannounce", 3, X.hivenumber)
+	var/list/allxenos = X.hive.get_all_xenos()
+	for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
+		sister.add_filter("summonoutline", 2, list("type" = "outline", "size" = 1, "color" = COLOR_VIOLET))
+
+	if(!do_after(X, 15 SECONDS, FALSE, X, BUSY_ICON_HOSTILE))
+		for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:124]: as()

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+			sister.remove_filter("summonoutline")
+		return fail_activate()
+
+	for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:128]: as()

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 6]
```diff
@@ -718,3 +718,34 @@
...
+			to_chat(owner, "<span class='warning'>Our hive does not have enough psychic energy to summon a King, we need [psych_cost] points!</span>")
+		return FALSE
+
+/datum/action/xeno_action/summon_king/action_activate()
+	var/mob/living/carbon/xenomorph/X = owner
+	to_chat(X, "<span class='xenonotice'>We begin constructing a psychic echo chamber for the Queen Mother...</span>")
+	if(!do_after(X, 15 SECONDS, FALSE, X, BUSY_ICON_HOSTILE))
+		return fail_activate()
+	SSpoints.xeno_points_by_hive[X.hivenumber] -= psych_cost
+	var/atom/pod = new /obj/structure/resin/king_pod(X.loc, X.hivenumber)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:748]: cast incorrectly

File: code/game/atoms.dm [hunk 7]
```diff
@@ -879,11 +879,13 @@ Proc for attack log creation, because really why not
  * originated_turf: if not null, will check if the obj_turf is closer than distance_max to originated_turf, and the proc will return if not
  * distance_max: used to check if originated_turf is close to obj.loc
 */
-/atom/proc/turn_light(mob/user = null, toggle_on , cooldown = 1 SECONDS, sparks = FALSE, forced = FALSE, originated_turf = null, distance_max = 0)
-	if(originated_turf && (get_dist(originated_turf, loc)<= distance_max))
+/atom/proc/turn_light(mob/user = null, toggle_on , cooldown = 1 SECONDS, sparks = FALSE, forced = FALSE, atom/originated_turf = null, distance_max = 0)
+	if(originated_turf && ((get_dist(originated_turf, loc) >= distance_max) || (originated_turf.z != loc.z)))
```
Comment (TiviPlus) [inline code/game/atoms.dm:883]: so is originated turf a turf or an atom?

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm (hunk 3 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:25]: the range check is done by turn light

Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:25]: it shouldnt be why is it

File: code/game/atoms.dm [hunk 8]
```diff
@@ -879,11 +879,11 @@ Proc for attack log creation, because really why not
  * originated_turf: if not null, will check if the obj_turf is closer than distance_max to originated_turf, and the proc will return if not
  * distance_max: used to check if originated_turf is close to obj.loc
 */
-/atom/proc/turn_light(mob/user = null, toggle_on , cooldown = 1 SECONDS, sparks = FALSE, forced = FALSE, originated_turf = null, distance_max = 0)
-	if(originated_turf && (get_dist(originated_turf, loc)<= distance_max))
-		return OUT_OF_REACH
```
[the remark this answers is not in the document]
Comment (BraveMole) [reply code/game/atoms.dm:884]: nop, removed

File: code/__DEFINES/mode.dm [hunk 9]
```diff
@@ -121,11 +121,14 @@
...
 #define XENO_TURRET_PRICE 150
+#define XENO_KING_PRICE 2500
+
+#define INVOKE_KING_TIME_LOCK 1 HOURS
 
 #define SMALL_SILO_MAXIMUM_PLAYER_COUNT 30
 
-//The minimum round time before siloless timer can start (13:00)
-#define MINIMUM_TIME_SILO_LESS_COLLAPSE 36000 
+//The minimum round time before siloless timer can start (13:03)
```
Comment (BraveMole) [inline code/__DEFINES/mode.dm:130]:
```suggestion
//Time (after round start) before siloless timer can start
```

**Comment by InterroLouis** [comment]
Why does King at Ancient have a spit delay var?

**Comment by David-Stormwell** [comment]
Can you give it a option to do so for xenos?/
"The king has summoned you, accept the call?"
Y/N

**Comment by Moggram** [comment]
You could have it in hunt If X number of bodies are in a King pool. Dead xenos also could be a used as a come back mechanic.

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+	succeed_activate()
+	add_cooldown()
+
+
+// ***************************************
+// *********** Gravity Crush
+// ***************************************
+#define ADD_FILTER (1<<0)
+#define REMOVE_FILTER (1<<1)
+#define GRAV_CRUSH (1<<2)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:36]: What is this why does it exist?

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+			if(action_flags & REMOVE_FILTER)
+				item.remove_filter("crushblur")
+			if(action_flags & GRAV_CRUSH)
+				if(isliving(item))
+					var/mob/living/mob_crushed = item
+					if(mob_crushed.stat == DEAD)//No abuse of that mechanic for some permadeath
+						continue
+					if(mob_crushed.faction == FACTION_XENO) //Don't harm the xeno. I thought it was better than a istype, could be wrong
+						continue
+				item.ex_act(EXPLODE_HEAVY)	//crushing without damaging the nearby area
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:96]: why does this add AND remove filters AND handle crushing

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+				item.add_filter("crushblur", 1, radial_blur_filter(0.3))
+				addtimer(CALLBACK(item, /atom.proc/remove_filter, "crushblur"), WINDUP_GRAV) //Fail safe if the item moves out of the zone.
+			if(action_flags & REMOVE_FILTER)
+				item.remove_filter("crushblur")
+			if(action_flags & GRAV_CRUSH)
+				if(isliving(item))
+					var/mob/living/mob_crushed = item
+					if(mob_crushed.stat == DEAD)//No abuse of that mechanic for some permadeath
+						continue
+					if(mob_crushed.faction == FACTION_XENO) //Don't harm the xeno. I thought it was better than a istype, could be wrong
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:94]: you are wrong this should be an owner hive check like I did it

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+/datum/action/xeno_action/activable/gravity_crush/proc/act_on_turfs(action_flags, list/turfs)
+	for(var/turf/targetted AS in turfs)
+		if(action_flags & ADD_FILTER)
+			targetted.add_filter("crushblur", 1, radial_blur_filter(0.3))
+		if(action_flags & REMOVE_FILTER)
+			targetted.remove_filter("crushblur")
+		for(var/atom/movable/item AS in targetted.contents)
+			if(action_flags & ADD_FILTER)
+				item.add_filter("crushblur", 1, radial_blur_filter(0.3))
+				addtimer(CALLBACK(item, /atom.proc/remove_filter, "crushblur"), WINDUP_GRAV) //Fail safe if the item moves out of the zone.
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:86]: timer spam

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm (hunk 11 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:96]: You said no copy paste?
I want to remove the filter AND crush as the same time

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+				continue
+			item.add_filter("crushblur", 1, list("type"="radial_blur", "size" = 0.3))
+		targetted.add_filter("crushblur", 1, list("type"="radial_blur", "size" = 0.3))
+	if(!do_after(owner, 2 SECONDS, FALSE, owner, BUSY_ICON_DANGER))
+		for(var/turf/targetted AS in turfs)
+			for(var/atom/movable/item AS in targetted.contents)
+				if(ismob(item))
+					continue
+				item.remove_filter("crushblur", 1, list("type"="radial_blur", "size" = 0.3))
+			targetted.remove_filter("crushblur")
```
[the remark this answers is not in the document]
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm]: this. I removed copypaste

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm (hunk 10 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:36]: the arguments to tell the act_on_crush proc what to do

**Comment by lKiseki** [comment]
As per this testmerge, APCs no longer kill lights.

Additionally, Nightfall seems to have no cooldown and no plasma cost, which means it can be spammed indefinitely.

**Comment by XSlayer300** [comment]
How about suitlights, they're still not affected, no?

**Comment by Isy232** [comment]
@BraveMole I dont want to be dumb here, but aren't all xenos supposed to be female?

I'm fine with all of the abilities the caste has, just having a male xenos seems a little out of character for the game.

**Comment by SplinterGP** [comment]
And for god sake if this goes in somehow remove it from hunt, it was completely awfull on it. You dont have stuff to use psy points on like turrets on it.

**Comment by BraveMole** [comment]
> Why are we adding king back with its same old abilities + 1 more, instead of something original with better thought,

How are kings ability not original? Nightfall ise literraly unique, grav crush doesn't look like anything (there are no winded up aoe ability like that, plus it warns the marine?), and the mass teleportation is unique as well?

And for the better thought, it's meant as a fob breaker, so it has two abilities that are meant for that (aoe to crush cades, and a tp to bypass defences). Plus another ability that's less cancer than the one before

> that thing was a balance hell and dindt fit in game any little and only served to disbalance game and give xenos a win when round took too long, it aint fun to play either, its point and click thing with 0 thought

Yeah and it changed, and the game with it? Grav crush was nerfed a lot, one ability was changed, the whole psy points system was redesigned, engie buffs, etc... 

For the fun it procures, i dunno, i didn't have the chance to play it, but unless you played with the current state of it you can't really say things like that?

**Comment by Losenis** [comment]
tonks when

**Comment by DeltaFire15** [comment]
a) This breaks lights for unpowered APCs.
Any APC that does not have power roundstart does gets active lights despite not being supposed to, which also do not turn off if the APC is destroyed. APCs that have power work fine (aka get slashed -> turn off)

b) Xeno King can turn on lights of disabled APCs, by affecting lights that got turned off earlier due to e.g. the APC getting broken, then having the ability run out which will reenable those lights despite being unpowered.

**Comment by JackTheJackhammer** [comment]
So, we're not getting tank with this?

What big boom cannon thing are we gonna kill this thing with? A SADAR?

Also Nightfall sounds like useless

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Im not merging this until it has wound sprites

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+	. = ..()
+	if(!.)
+		return
+	if(!owner.line_of_sight(A, king_crush_dist))
+		if(!silent)
+			to_chat(owner, "<span class='warning'>We must get closer to crush, our mind cannot reach this far.</span>")
+		return FALSE
+
+/datum/action/xeno_action/activable/gravity_crush/use_ability(atom/A)
+	var/list/turfs = RANGE_TURFS(1, A)
```
Comment (Wayland-Smithy) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:63]:
```suggestion
/datum/action/xeno_action/activable/gravity_crush/use_ability(atom/A)
	owner.face_atom(A) //Face towards the target so we don't look silly
	var/list/turfs = RANGE_TURFS(1, A)
```

Comment (Wayland-Smithy) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:63]:
```suggestion
	owner.face_atom(A)
	var/list/turfs = RANGE_TURFS(1, A)
```

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**